### DISCUSSION TITLE
How to improve Performance for student grouping algorithm based on BalanceGroupSat

### DISCUSSION THREAD

**Discussion by zdz72113** [google/or-tools#4091] [CP-SAT questions] [answered]
**What version of OR-Tools and what language are you using?**
Version: v9.5.
Language: C#

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
CP-SAT

**What did you do?**
I have implemented a student grouping algorithm based on the example code at https://github.com/google/or-tools/blob/stable/examples/dotnet/BalanceGroupSat.cs. my algorithm aims to assign students to classes with equal numbers, equal male-female ratio, and with the objective of the total score and the subject scores of every class as close as possible to their average of all students.

I have tested the algorithm with 200 students data, and it takes only 0.29 seconds to obtain the optimal result when dividing the students into 2 groups. However, when dividing them into 10 groups, the algorithm runs for 4 minutes without reaching an optimal result.

attached the source code, logs from two test runs, and the format of the test data. Could you please provide guidance on how to optimize the algorithm's performance? Thank you.

[source_code.txt](https://github.com/google/or-tools/files/14331462/source_code.txt)
[log_200_items_in_2_groups.txt](https://github.com/google/or-tools/files/14331451/log_200_items_in_2_groups.txt)
[log_200_items_in_10_groups.txt](https://github.com/google/or-tools/files/14331452/log_200_items_in_10_groups.txt)
![test_data](https://github.com/google/or-tools/assets/14866083/3845fbc5-6c9e-4daf-8087-97f23131e18e)

**Comment by zdz72113** [comment]
![source_code](https://github.com/google/or-tools/assets/14866083/2227051f-965b-4eda-a0e0-4692f639f6e0)

**Comment by lperron** [comment, marked as answer]
logs are not in ascii format.
Please run model.ExportToFile('2groups.pb.txt') (or 10groups.pb.txt) and attach the proto files.

Try to upgrade to 9.8 and use more workers.

**Reply by zdz72113** [reply]
thanks, finally the issue was solved via iterative calculations.
